source/kernel/src/registry/mod.rs

```diff
@@ -709,38 +816,56 @@ impl Registry {
                 // Safety: `handshake_result` is guaranteed to be initialized by
                 // `erased_handshake` if and only if its future completes with
                 // an `Ok(())`. and it did!
-                handshake_result
-                    .assume_init()
-                    .map_err(UserConnectError::Connect)?
-                    .type_erase()
+                handshake_result.assume_init()?.type_erase()
             },
         };
 
         let client_id = self.counter.fetch_add(1, Ordering::Relaxed);
         info!(
             svc = %any::type_name::<RD>(),
             uuid = ?RD::UUID,
-            service_id = item.value.service_id.0,
+            service_id = service_id.0,
             client_id,
             "Got KernelHandle from Registry",
         );
 
         Ok(UserspaceHandle {
             req_producer_leaked,
             req_deser: vtable.req_deser,
-            service_id: item.value.service_id,
+            service_id,
             client_id: ClientId(client_id),
         })
     }
 
-    fn get<RD: RegisteredDriver>(&self) -> Result<&RegistryItem, ConnectError<RD>> {
-        let Some(item) = self.items.as_slice().iter().find(|i| i.key == RD::UUID) else {
+    async fn insert_item(&self, item: RegistryItem) -> Result<(), RegistrationError> {
+        {
+            let mut items = self.items.write().await;
```

Comment:
Add "load bearing scope" notice here too?

Reply:
i don't actually think this one is load-bearing --- it's equivalent anyway, since the woken tasks won't get to run until this function returns, and the lock is dropped at the end of the function. this only makes things slightly more efficient in concurrent use-cases, where the wake call might wake a task on another core that might try to grab the lock. but this can't deadlock in either case.